add use_fake_all_gather and use_fake_reduce_scatter to FSDP for ablation studies

ckluk2 opened the pull request:
Summary: As titled

Test Plan: Not needed because this is only for doing ablation studies

Reviewed By: awgu

Differential Revision: D50867908




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang @voznesenskym @penguinwu @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

wanchaol commented on torch/distributed/fsdp/_flat_param.py:
```diff
@@ -95,6 +95,13 @@
+# communication ops for ablation studies. Note that without these communication
+# ops the training won't converge, and you probably need to disable correctness
+# checks in your model.
+_FSDP_USE_FAKE_ALL_GATHER = "FSDP_USE_FAKE_ALL_GATHER"
```
instead of doing this, I am wondering if you should just make a FakeProcessGroup instead? https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/distributed/fake_pg.py#L25

ckluk2 replied:
Thanks for the suggestion! I didn't know the existence of the FakeProcessGroup. Let me try it.

awgu replied:
My bad. I forgot about `FakeProcessGroup`.

wanchaol replied:
Some reference test example in case it's easier to play with it https://github.com/pytorch/pytorch/blob/main/test/distributed/test_fake_pg.py#L82

ckluk2 replied:
I change to use FakeProcessGroup, as suggested by wanchaol.

awgu commented on torch/distributed/fsdp/_flat_param.py:
```diff
@@ -490,19 +498,39 @@ def __init__(
         self._device_handle = _FSDPDeviceHandle.from_device(self.device)
         self.process_group = process_group
+        if self._use_fake_all_gather or self._use_fake_reduce:
+            self._fake_process_group = FakeProcessGroup(
```
cc: @wanchaol Is there any concern of constructing 1 `FakeProcessGroup` for every FSDP instance (instead of reusing a single one)? It seems okay?

wanchaol replied:
It seems ok to reconstruct FakePG if needed. I am mainly wondering why we choose to create a FakePG inside FSDP, rather than create a FakePG, and pass it to FSDP? the latter approach don't need any FSDP specific changes IIUC

awgu replied:
Good point. @ckluk2 Can you try doing that?

(pass a `FakeProcessGroup` directly to FSDP's arg)
https://github.com/pytorch/pytorch/blob/9062e429db5bcba306a7ef5e1d6582bef564bca2/torch/distributed/fsdp/fully_sharded_data_parallel.py#L413

ckluk2 replied:
* I actually prefer adding this feature to FSDP, because users would like to use this feature without changing this code
* Also, we will need to do some FSDP-specific change anyway, because we want to separately measure the performance impact of all_gather and reduce_scatter.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -914,7 +919,12 @@ def _reduce_grad_no_shard(state: _FSDPState, handle: FlatParamHandle) -> None:
     if state._comm_hook is None:  # default path
         _div_if_needed(flat_param.grad, state._gradient_predivide_factor)
-        dist.all_reduce(flat_param.grad, group=state.process_group)
+        pg = (
+            handle._fake_process_group
+            if handle._use_fake_reduce
+            else state.process_group
+        )
+        dist.all_reduce(flat_param.grad, group=pg)
```
Just to check: This is for `NO_SHARD`'s all-reduce. Do we want to fakify this?

awgu commented on torch/distributed/fsdp/_flat_param.py:
```diff
@@ -1480,9 +1510,13 @@ def unshard_grad(self):
-        dist.all_gather_into_tensor(
-            padded_unsharded_grad, sharded_grad, self.process_group
+
+        pg = (
+            self._fake_process_group
+            if self._use_fake_all_gather
+            else self.process_group
         )
```
This `unshard_grad` should only be called from `summon_full_params(with_grads=True)`, which is not normally called during training. I am not sure we have to fakify this one.

ckluk2 replied:
Don't fakify this one too.

awgu reviewed (approved):
Sounds good!

ckluk2 committed:
add use_fake_all_gather and use_fake_reduce_scatter to FSDP for ablation studies (#113106)

Summary:
As titled


Test Plan: Not needed because this is only for doing ablation studies

Differential Revision: D50867908

Pulled By: ckluk2

ckluk2 commented:
@pytorchbot merge
